### PR TITLE
[QNN EP] Reduce attention mask value

### PR DESCRIPTION
**Pull request by tzuhsuanwei**
## Summary

Recognize a QDQ attention tail in the QNN node group and reduce the mask value while preserving the original model structure.

## Validation

Validated on a V81 device with the target quantized attention model across 566 cases.

### CODE REVIEW

File: onnxruntime/core/providers/qnn/builder/qnn_model_wrapper.h [hunk 1]
```diff
@@ -600,6 +622,10 @@ class QnnModelWrapper {
   std::unordered_map<std::string, QnnTensorWrapper> model_tensors_map_;
   // All QnnParamWrapper for the graph
   std::unordered_map<std::string, QnnParamWrapper> model_params_map_;
+  // Quantization overrides installed by node-groups before an OpBuilder emits
+  // the corresponding output tensor.
+  std::unordered_map<std::string, QnnQuantParamsWrapper> output_quant_param_overrides_;
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_model_wrapper.h]: Is it really necessary having two containers? The tensor name should be unique.

File: onnxruntime/core/providers/qnn/builder/qnn_model_wrapper.cc [hunk 2]
```diff
@@ -83,6 +83,9 @@ Ort::Status QnnModelWrapper::MakeTensorWrapper(const OrtNodeUnitIODef& tensor, Q
 
   TensorInfo tensor_info = {};
   RETURN_IF_ERROR(GetTensorInfo(tensor, tensor_info));
+  if (const auto* quant_override = GetTensorQuantParamOverride(tensor_name); quant_override != nullptr) {
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_model_wrapper.cc]: Can the override happen in `GetTensorInfo`?

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+// Recognized PSL tail (all values are QDQ-wrapped):
+//
+//   attention_mask -> Sub(1, mask) -> Mul(mask, -10000) --+
+//   score ---------> Div ----------------------------------> Add -> Add(gate) -> Softmax
+//
+// The lowering keeps this topology and changes the mask value to -100.
+class FixMaskValueFusion final : public IQnnNodeGroup {
+ public:
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(FixMaskValueFusion);
+  ~FixMaskValueFusion() override;
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.h]: It seems redundant to override.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  static std::unique_ptr<IQnnNodeGroup> TryFusion(
+      QnnModelWrapper& qnn_model_wrapper,
+      const OrtNodeUnit& softmax_node_unit,
+      const std::unordered_map<const OrtNode*, const OrtNodeUnit*>& node_to_node_unit,
+      const std::unordered_map<const OrtNodeUnit*, const IQnnNodeGroup*>& node_unit_to_qnn_node_group,
+      const Ort::Logger& logger);
+
+ private:
+  struct Match;
+  explicit FixMaskValueFusion(std::unique_ptr<Match> match);
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.h]: Align with others to make it public.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+
+#include "core/providers/qnn/builder/qnn_node_group/qnn_node_group.h"
+
+namespace onnxruntime::qnn {
+
+// Recognized PSL tail (all values are QDQ-wrapped):
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.h]: Remove the model naming, and as well for others.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+#include "core/providers/qnn/builder/qnn_model_wrapper.h"
+#include "core/providers/qnn/builder/qnn_node_group/utils.h"
+
+namespace onnxruntime::qnn {
+
+constexpr double kOriginalMaskMagnitude = 10000.0;
+constexpr double kReplacementMaskMagnitude = 100.0;
+constexpr double kMaskScaleRatio = kReplacementMaskMagnitude / kOriginalMaskMagnitude;
+
+struct FixMaskValueFusion::Match {
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: Since there may be other pattern in the future, the naming should be able to extend.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+bool IsRank4(const OrtNodeUnitIODef& io) {
+  return io.shape.has_value() && io.shape->size() == 4;
+}
+
+bool SameShape(const OrtNodeUnitIODef& lhs, const OrtNodeUnitIODef& rhs) {
+  return lhs.shape.has_value() && rhs.shape.has_value() && *lhs.shape == *rhs.shape;
+}
+
+template <typename Match>
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: Why the function needs template?

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+bool MatchPslTail(const QnnModelWrapper& qmw, const OrtNodeUnit& softmax,
+                  const std::unordered_map<const OrtNode*, const OrtNodeUnit*>& node_map,
+                  const std::unordered_map<const OrtNodeUnit*, const IQnnNodeGroup*>& claimed,
+                  Match& match) {
+  match = {};
+  if (softmax.OpType() != "Softmax" || softmax.Inputs().size() != 1 || softmax.Outputs().size() != 1 ||
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: I don't think Softmax would ever have more than 1 input / 1 output / non-matched IO shape.
Reduce the check (as well for others) to safely assume nodes are following ONNX spec, and we may skip those checks.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+  const int64_t axis = attrs.Get("axis", static_cast<int64_t>(-1));
+  if (axis != -1 && axis != 3) return false;
+
+  const OrtNodeUnit* gated_add = ParentForInput(qmw, softmax, 0, node_map, claimed);
+  if (!IsBinary(gated_add, "Add")) return false;
+  const OrtNodeUnit* lhs = ParentForInput(qmw, *gated_add, 0, node_map, claimed);
+  const OrtNodeUnit* rhs = ParentForInput(qmw, *gated_add, 1, node_map, claimed);
+  const OrtNodeUnit* masked_add = nullptr;
+  const OrtNodeUnit* gate = nullptr;
+  if (IsBinary(lhs, "Add") && IsBinary(rhs, "Mul")) {
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: I suppose only one of these two branches exists in the model.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  match = {};
+  if (softmax.OpType() != "Softmax" || softmax.Inputs().size() != 1 || softmax.Outputs().size() != 1 ||
+      !IsRank4(softmax.Inputs()[0]) || !SameShape(softmax.Inputs()[0], softmax.Outputs()[0])) {
+    return false;
+  }
+  const OrtNodeAttrHelper attrs(softmax);
+  const int64_t axis = attrs.Get("axis", static_cast<int64_t>(-1));
+  if (axis != -1 && axis != 3) return false;
+
+  const OrtNodeUnit* gated_add = ParentForInput(qmw, softmax, 0, node_map, claimed);
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: Comment which part of the pattern that each block is checking.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+  if (mask_sub == nullptr || !IsRank4(mask_sub->Outputs()[0]) ||
+      !IsStaticQdqInputWithValue(qmw, *mask_sub, 0, node_map, 1.0, /*require_scalar=*/true) ||
+      qmw.IsConstantInput(mask_sub->Inputs()[1].name) || !IsRank4(mask_sub->Inputs()[1])) {
+    return false;
+  }
+  const size_t mask_bias_index = mul_lhs == mask_sub ? 1 : 0;
+  const OrtNodeUnitIODef& mask_bias_input = mask_mul->Inputs()[mask_bias_index];
+  // Require PSL's static -10000 bias by value, not exporter tensor name.
+  if (!IsStaticQdqInputWithValue(qmw, *mask_mul, mask_bias_index, node_map, -kOriginalMaskMagnitude,
+                                 /*require_scalar=*/false)) {
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: Why no need to be scalar?

File: onnxruntime/test/providers/qnn/qnn_node_group/fix_mask_value_fusion_test.cc [hunk 12]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (minfhong-qti) [inline onnxruntime/test/providers/qnn/qnn_node_group/fix_mask_value_fusion_test.cc]: Check again what test case should be wriiten.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+    const std::unordered_map<const OrtNode*, const OrtNodeUnit*>& node_to_node_unit,
+    const std::unordered_map<const OrtNodeUnit*, const IQnnNodeGroup*>& claimed,
+    [[maybe_unused]] const Ort::Logger& logger) {
+  std::unique_ptr<Match> match(new Match());
+  if (!MatchPslTail(qmw, softmax_node_unit, node_to_node_unit, claimed, *match)) return nullptr;
+  return std::unique_ptr<IQnnNodeGroup>(new FixMaskValueFusion(std::move(match)));
+}
+
+FixMaskValueFusion::FixMaskValueFusion(std::unique_ptr<Match> match) : match_(std::move(match)) {
+  match_->node_units = {match_->masked_add, match_->gated_add, match_->softmax};
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: Can `node_units` be derived inside the struct?

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+struct FixMaskValueFusion::Match {
+  const OrtNodeUnit* softmax = nullptr;
+  const OrtNodeUnit* gated_add = nullptr;
+  const OrtNodeUnit* masked_add = nullptr;
+  const OrtNodeUnit* score = nullptr;
+  const OrtNodeUnit* gate = nullptr;
+  const OrtNodeUnit* mask_mul = nullptr;
+  const OrtNodeUnit* mask_sub = nullptr;
+  const OrtNodeUnitIODef* mask_bias = nullptr;
+  bool claims_legacy_mask_chain = false;
```
Comment (minfhong-qti) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/fix_mask_value_fusion.cc]: Comment why this flag is required.
